File: server/regions/regions.go
package regions

import (
	"fmt"
	"log"
	"strconv"
	"strings"
	"time"

	"legend-of-adventure/server/entities"
	"legend-of-adventure/server/events"
	"legend-of-adventure/server/terrain"
)

// Region IDs should be structured like this:
//     parent:type x y
// parent: The ID of the parent or the name of the world
// type: The type of area that the region is
// x: The X coordinate of the area
// y: The Y coordinate of the area
// Examples:
//     overworld,field:0:0
//     overworld,field:0:0,house0 1 0
//     overworld,field:123:456,dungeon:0:0
//     overworld,field:123:456,dungeon:2:3,dungeon:1:1
//     overworld,field:123:456,dungeon:2:3,dungeon:1:1


type regionRequest struct {
	ID       string
	Receiver chan *Region
}

var regionCache = make(map[string]*Region)
var regionGetter = make(chan regionRequest, 64)
var startedRegionGetter = false

func startRegionGetter() {
	startedRegionGetter = true
	go func() {
		for {
			select {
			case request := <-regionGetter:
				parent, regType, x, y := GetRegionData(request.ID)
				reg := new(Region)
				reg.ParentID = parent
				reg.Type = regType
				reg.X = x
				reg.Y = y
				reg.killer = make(chan bool, 1)
				reg.doTTL()

				reg.entities = make([]*entities.Entity, 0, 32)
				reg.Terrain = terrain.Get(reg)

				if reg.IsTown() {
					terrain.ApplyTown(reg.Terrain)

				} else if reg.IsDungeonEntrance() {
					terrain.ApplyDungeonEntrance(reg.Terrain)

				} else if reg.Type == terrain.REGIONTYPE_DUNGEON {
					terrain.ApplyDungeon(parent, reg.Terrain)

				} else if terrain.IsBuildingType(reg.Type) {
					terrain.ApplyBuildingInterior(reg.Terrain, reg.Type, parent)
				}

				reg.PopulateEntities();

				regionCache[request.ID] = reg

				request.Receiver <- reg
			}
		}
	}()
}

func getRegionID(parent, regType string, x, y int) string {
	return parent + "," + regType + ":" + strconv.Itoa(x) + ":" + strconv.Itoa(y)
}
func GetRegionData(ID string) (string, string, int, int) {
	parentSplit := strings.Split(ID, ",")
	if len(parentSplit) < 2 {
		return terrain.WORLD_OVERWORLD, terrain.REGIONTYPE_FIELD, 0, 0
	}
	regSplit := strings.Split(parentSplit[len(parentSplit) - 1], ":")
	if len(regSplit) != 3 {
		return terrain.WORLD_OVERWORLD, terrain.REGIONTYPE_FIELD, 0, 0
	}
	parent := strings.Join(parentSplit[:len(parentSplit) - 1], ",")
	x, _ := strconv.ParseInt(regSplit[1], 10, 0)
	y, _ := strconv.ParseInt(regSplit[2], 10, 0)
	return parent, regSplit[0], int(x), int(y)
}

func GetRegion(parent, regType string, x, y int) *Region {
	if !startedRegionGetter {
		startRegionGetter()
	}

	regionID := getRegionID(parent, regType, x, y)

	if !IsValidRegionID(regionID) {
		log.Println("Invalid region ID requested: " + regionID)
		return nil
	}

	if reg, ok := regionCache[regionID]; ok {
		return reg
	}

	responseChan := make(chan *Region, 1)
	request := regionRequest{
		regionID,
		responseChan,
	}
	regionGetter <- request

	return <-responseChan
}

type Region struct {
	ParentID string
	Type     string
	X, Y     int

	// Bits and pieces to clean up the region.
	KeepAlive chan bool
	killer    chan bool

	Terrain  *terrain.Terrain
	entities []*entities.Entity
}

func (self *Region) Broadcast(evt *events.Event) {
	for _, entity := range self.entities {
		if (*entity).ID() == evt.Origin {
			continue
		}
		(*entity).Receive() <- evt
	}
}

func (self *Region) doTTL() {
	self.KeepAlive = make(chan bool)
	go func(self *Region) {
		for {
			select {
			case <-self.KeepAlive:
				log.Println("Keeping region " + self.ID() + " alive.")

			case <-time.After(2 * time.Minute):
				log.Println("Region " + self.ID() + " timed out.")
				// Remove references to the region from the region cache.
				delete(regionCache, self.ID())
				// Tell the entities that are listening that they need to clean up.
				self.killer <- true
				close(self.KeepAlive)

				return
			}
		}
	}(self)
}

func (self Region) ID() string {
	return getRegionID(self.ParentID, self.Type, self.X, self.Y)
}

func (self *Region) GetEvent(evt_type events.EventType, body string, origin entities.Entity) *events.Event {
	str_origin := ""
	if origin != nil {
		str_origin = origin.ID()
	}

	return &events.Event{self.ID(), evt_type, str_origin, body}
}

func (self *Region) AddEntity(entity entities.Entity) {
	entity.Killer(self.killer)

	// Add the entity to the list of entities.
	self.entities = append(self.entities, &entity)

	// Tell everyone else that the entity is here.
	x, y := entity.BlockingPosition()
	self.Broadcast(
		self.GetEvent(
			events.REGION_ENTRANCE,
			fmt.Sprintf(
				"%s\n%f %f",
				entity.BlockingString(),
				x, y,
			),
			entity,
		),
	)

	// Tell the entity about everyone else.
	for _, regEnt := range self.entities {
		if regEnt == &entity { continue }

		x, y := (*regEnt).BlockingPosition()
		entity.Receive() <- self.GetEvent(
			events.REGION_ENTRANCE,
			fmt.Sprintf(
				"%s\n%f %f",
				<-((*regEnt).String()),
				x, y,
			),
			*regEnt,
		)
	}

}

func (self *Region) RemoveEntity(entity entities.Entity) {
	entity.Killer(self.killer)

	// Tell everyone else that the entity is leaving.
	self.Broadcast(self.GetEvent(events.REGION_EXIT, entity.ID(), entity))

	// Find the entity
	location := -1
	for i, e := range self.entities {
		if *e == entity {
			location = i
			break
		}
	}

	if location == -1 {
		log.Println("Could not find entity to remove!")
		return
	}

	// ...and remove it
	self.entities = append(self.entities[:location], self.entities[location+1:]...)

}

func (self *Region) String() string {
	tileset := terrain.GetTileset(self.GetRoot(), self.GetType())
	return self.Terrain.String() + ", \"tileset\": \"" + tileset + "\", \"can_slide\": true"
}

func (self Region) GetRoot() string {
	parent := self.ParentID
	for parent != terrain.WORLD_OVERWORLD && parent != terrain.WORLD_ETHER {
		parent, _, _, _ = GetRegionData(parent)
	}
	return parent
}

func (self Region) GetTerrain() *terrain.Terrain {
	return self.Terrain
}

func (self Region) GetParent() string {
	return self.ParentID
}

func (self Region) GetType() string {
	return self.Type
}

func (self Region) GetX() int {
	return self.X
}

func (self Region) GetY() int {
	return self.Y
}

func (self Region) IsTown() bool {
	if self.ParentID != terrain.WORLD_OVERWORLD &&
		self.ParentID != terrain.WORLD_ETHER &&
		self.Type != terrain.REGIONTYPE_FIELD {
		return false
	}
	return isTownPos(self.X, self.Y)
}

func (self Region) GetEntity(ID string) entities.Entity {
	for _, entity := range self.entities {
		if (*entity).ID() == ID {
			return *entity
		}
	}
	return nil
}

func (self *Region) PopulateEntities() {
	rng := terrain.GetCoordRNG(float64(self.X), float64(self.Y))

	placeEntity := func(entType string) {
		// TODO: Find a way to figure this out.
		entW, entH := 1.0, 1.0
		for {
			x := rng.Float64() * float64(self.Terrain.Width - 2 - uint(entW)) + 1
			y := rng.Float64() * float64(self.Terrain.Height - 2 - uint(entH)) + 1 + entH
			if !self.Terrain.Hitmap.Fits(x, y, entW, entH) { continue }
			self.Spawn(entType, x, y)
			return
		}
	}

	switch self.Type {
	case terrain.REGIONTYPE_FIELD:
		entCount := rng.Intn(MAX_ENTITIES_PER_FIELD)
		for i := 0; i < entCount; i++ {
			var entType string
			if i % WOLF_ODDS == 0 {
				entType = "wolf"
			} else {
				entType = "sheep"
			}
			placeEntity(entType)
		}

		if self.IsTown() {
			soldierCount := rng.Intn(MAX_SOLDIERS_PER_TOWN - MIN_SOLDIERS_PER_TOWN) + MIN_SOLDIERS_PER_TOWN
			for i := 0; i < soldierCount; i++ {
				placeEntity("soldier")
			}

			placeEntity("bully")
			placeEntity("child")
			placeEntity("child")
			placeEntity("child")

			// placeEntity("trader")

			// self.Spawn("test", 50, 50)
		}

	case terrain.REGIONTYPE_SHOP:
		placeEntity("homely")
		placeEntity("homely")

		fallthrough
	case terrain.REGIONTYPE_HOUSE:
		placeEntity("homely")
		placeEntity("homely")

		if rng.Intn(SOLDIER_IN_HOUSE_ODDS) == 0 {
			placeEntity("soldier")
		}
		if rng.Intn(TRADER_IN_HOUSE_ODDS) == 0 {
			placeEntity("trader")
		}

		totalTiles := self.Terrain.Width * self.Terrain.Height
		for i := uint(0); i < totalTiles; i++ {
			tile := self.Terrain.Tiles[i % self.Terrain.Height][i / self.Terrain.Width]
			if tile != 58 { continue }

			chest := entities.NewChestEntity(self, float64(i / self.Terrain.Width), float64(i % self.Terrain.Height) + 1.25)
			self.AddEntity(chest)

			items := rng.Intn(MAX_ITEMS_PER_SHOP_CHEST - MIN_ITEMS_PER_SHOP_CHEST) + MIN_ITEMS_PER_SHOP_CHEST
			for j := 0; j < items; j++ {
				var code string
				if rng.Intn(10) < ODDS_SHOP_CHEST_SWORD {
					code = fmt.Sprintf(
						"wsw.%s.%d",
						entities.WEAPON_RAW_PREFIXES[rng.Intn(len(entities.WEAPON_RAW_PREFIXES))],
						rng.Intn(SHOP_CHEST_SWORD_MAX_LEV),
					)
				} else {
					var idx int
					if rng.Intn(10) < ODDS_SHOP_CHEST_POTION {
						code = "p"
						idx = rng.Intn(10)
					} else {
						code = "f"
						idx = rng.Intn(9)
					}
					code = fmt.Sprintf("%s%d", code, idx)
				}
				log.Println("Adding " + code + " to a chest")
				chest.AddItem(code)
			}
		}

	case terrain.REGIONTYPE_DUNGEON:
		entCount := rng.Intn(MAX_ENTITIES_PER_DUNGEON)
		for i := 0; i < entCount; i++ {
			var entType string
			if i % DEATH_WAKER_ODDS == 0 {
				entType = "zombie"
			} else {
				entType = "death_waker"
			}
			placeEntity(entType)
		}
	}

}

func (self *Region) Spawn(entType string, x, y float64) {
	ent := entities.NewVirtualEntity(entType)
	ent.SetLocation(self)
	ent.SetPosition(x, y)
	self.AddEntity(ent)
}


// Helper Methods:

func isTownPos(x, y int) bool {
	// Always force spawn to be a town.
	return x == 0 && y == 0 || terrain.GetCoordOption(x, y, ODDS_TOWN)
}

func (self Region) IsDungeonEntrance() bool {
	if self.ParentID != terrain.WORLD_OVERWORLD && self.ParentID != terrain.WORLD_ETHER {
		return false
	}
	return isDungeonPos(self.X, self.Y)
}

func isDungeonPos(x, y int) bool {
	// Always force 1 0 to be a dungeon.
	return x == 1 && y == 0 || terrain.GetCoordOption(x, y, ODDS_DUNGEON)
}

func IsValidRegionID(ID string) bool {
	parent, regType, _, _ := GetRegionData(ID)
	if parent == terrain.WORLD_OVERWORLD || parent == terrain.WORLD_ETHER {
		return regType == terrain.REGIONTYPE_FIELD
	}

	if !IsValidRegionID(parent) {
		return false
	}

	_, parentType, parentX, parentY := GetRegionData(parent)

	if regType == terrain.REGIONTYPE_DUNGEON {
		return (parentType == terrain.REGIONTYPE_DUNGEON ||
			parentType == terrain.REGIONTYPE_FIELD && isDungeonPos(parentX, parentY))
	}
	if regType == terrain.REGIONTYPE_FIELD {
		return parentType == terrain.WORLD_OVERWORLD || parentType == terrain.WORLD_ETHER
	}

	return true
}

File: src/regions.js
const entity = require('./entity');
const events = require('./events');
const player = require('./player');
const terrain = require('./terrain');


const DEFAULT_REGION_DATA = [terrain.WORLD_OVERWORLD, terrain.REGIONTYPE_FIELD, 0, 0];
const ODDS_DUNGEON = 14;
const ODDS_TOWN = 14;
const CLEANUP_TTL = 60 * 1000;

const MAX_ENTITIES_PER_FIELD = 5;
const MAX_ENTITIES_PER_DUNGEON = 4;
const MAX_SOLDIERS_PER_TOWN = 3;
const MIN_SOLDIERS_PER_TOWN = 1;
const SOLDIER_IN_HOUSE_ODDS = 2;
const TRADER_IN_HOUSE_ODDS = 4;
const WOLF_ODDS = 4;
const DEATH_WAKER_ODDS = 4;

const MAX_ITEMS_PER_SHOP_CHEST = 10;
const MIN_ITEMS_PER_SHOP_CHEST = 1;
const ODDS_SHOP_CHEST_EMPTY = 2;  // out of 10
const ODDS_SHOP_CHEST_POTION = 2; // out of 10
const ODDS_SHOP_CHEST_SWORD = 2;  // out of 10
const ODDS_SHOP_CHEST_GUARD = 1;  // out of 10
const SHOP_CHEST_SWORD_MAX_LEV = 10;

const TICK_INTERVAL = 100;

const regionCache = {};


class Region {
  constructor(parent, type, x, y) {
    this.parentID = parent;
    this.type = type;
    this.x = x;
    this.y = y;

    this.id = getRegionID(parent, type, x, y);
    regionCache[this.id] = this;

    this.terrain = new terrain.Terrain(this);

    this.entities = new Set();
    this.entityMap = new Map();

    this.populateEntities();

    this.cleanup = null;

    this.ticker = setInterval(this.tick.bind(this), TICK_INTERVAL);
  }

  prepForDeletion() {
    this.cleanup = setTimeout(() => {
      delete regionCache[this.id];
      clearInterval(this.ticker);
    }, CLEANUP_TTL);
  }

  tick() {
    let hasPlayers = false;
    for (let entity of this.entities.values()) {
      if (entity instanceof player.Player) {
        hasPlayers = true;
      }
      entity.tick();
    }

    if (!hasPlayers && !this.cleanup) {
      this.prepForDeletion();
    }
  }

  broadcast(event) {
    for (let entity of this.entities) {
      if (entity === event.origin) {
        continue;
      }
      entity.onEvent(event);
    }
  }

  isDungeonEntrance() {
    if (this.parentID !== terrain.WORLD_OVERWORLD && this.parentID !== terrain.WORLD_ETHER) {
      return false;
    }

    return isDungeonPos(this.x, this.y);
  }
  isTown() {
    if (
      this.parentID !== terrain.WORLD_OVERWORLD &&
      this.parentID !== terrain.WORLD_ETHER &&
      this.type !== terrain.REGIONTYPE_FIELD
    ) {
      return false;
    }

    return isTownPos(this.x, this.y);
  }
  populateEntities() {
    const rng = terrain.getCoordRNG(this.x, this.y);

    const placeEntity = entType => {
      let entW = 1;
      let entH = 1;
      while (true) {
        const x = rng.uniform() * (this.terrain.width - 2 - entW) + 1;
        const y = rng.uniform() * (this.terrain.height - 2 - entH) + 1;
        if (!this.terrain.hitmap.fits(x, y, entW, entH)) {
          continue;
        }
        this.spawn(entType, x, y);
        break;
      }
    };

    switch (this.type) {
      case terrain.REGIONTYPE_FIELD:
        this.populateFieldEntities(rng, placeEntity);
        break;

      case terrain.REGIONTYPE_SHOP:
        placeEntity('homely');
        placeEntity('homely');

      case terrain.REGIONTYPE_HOUSE:
        placeEntity('homely');
        placeEntity('homely');
        this.populateHouseEntities(rng, placeEntity);
        break;

      case terrain.REGIONTYPE_DUNGEON:
        this.populateDungeonEntities(rng, placeEntity);
        break;
    }
  }

  populateFieldEntities(rng, placeEntity) {
    const entCount = rng.uniform() * MAX_ENTITIES_PER_FIELD;
    for (let i = 0; i < entCount; i++) {
      placeEntity(i % WOLF_ODDS === 0 ? 'wolf' : 'sheep');
    }

    if (this.isTown()) {
      const soldierCount = rng.range(MIN_SOLDIERS_PER_TOWN, MAX_SOLDIERS_PER_TOWN);
      for (let i = 0; i < soldierCount; i++) {
        placeEntity('soldier');
      }

      placeEntity('bully');
      placeEntity('child');
      placeEntity('child');
      placeEntity('child');

      placeEntity('trader');
    }
  }

  populateHouseEntities(rng, placeEntity) {
    if (rng.range(0, SOLDIER_IN_HOUSE_ODDS) === 0) {
      placeEntity('soldier');
    }
    if (rng.range(0, TRADER_IN_HOUSE_ODDS) === 0) {
      placeEntity('trader');
    }

    const totalTiles = this.terrain.width * this.terrain.height;

    for (let i = 0; i < totalTiles; i++) {
      const tile = this.terrain.tiles[i % this.terrain.height][i / this.terrain.width];
      if (tile === 58) {
        this.placeChestShop(
          i / this.terrain.width,
          i % this.terrain.height + 0.75,
          rng
        );
      } else if (tile === 59) {
        this.placePotShop(
          i / this.terrain.width,
          i % this.terrain.height + 1,
          rng
        );
      }
    }
  }

  populateDungeonEntities(rng, placeEntity) {
    const entCount = rng.range(0, MAX_ENTITIES_PER_DUNGEON);
    for (let i = 0; i < entCount; i++) {
      placeEntity(i % DEATH_WAKER_ODDS === 0 ? 'death_waker' : 'zombie');
    }
  }

  placeChestShop(x, y, rng) {
    const chest = new entity.ChestEntity(this, x, y);
    this.addEntity(chest);

    const items = rng.range(MIN_ITEMS_PER_SHOP_CHEST, MAX_ITEMS_PER_SHOP_CHEST);
    for (let i = 0; i < items; i++) {
      let code;
      if (rng.range(0, 10) < ODDS_SHOP_CHEST_SWORD) {
        code = `wsw.${
          entity.WEAPON_RAW_PREFIXES[rng.range(0, entities.WEAPON_RAW_PREFIXES.length - 1)]
        }.${
          rng.range(0, SHOP_CHEST_SWORD_MAX_LEV)
        }`;

      } else {
        if (rng.range(0, 10) < ODDS_SHOP_CHEST_POTION) {
          code = `p${rng.range(0, 10)}`;
        } else {
          code = `f${rng.range(0, 9)}`;
        }
      }
      chest.addItem(code);
    }
  }
  placePotShop(x, y, rng) {
    const pot = new entity.PotEntity(this, x, y);
    this.addEntity(pot);

    if (rng.rnage(0, 10) < ODDS_SHOP_CHEST_EMPTY) {
      return;

    } else if (rng.range(0, 10) < ODDS_SHOP_CHEST_GUARD) {
      pot.addEntity('soldier');
      return;

    } else if (rng.range(0, 10) < ODDS_SHOP_CHEST_SWORD) {
      pot.addItem(
        `wsw.${
          entity.WEAPON_RAW_PREFIXES[rng.range(0, entities.WEAPON_RAW_PREFIXES.length - 1)]
        }.${
          rng.range(0, SHOP_CHEST_SWORD_MAX_LEV)
        }`
      );

    } else {
      if (rng.range(0, 10) < ODDS_SHOP_CHEST_POTION) {
        pot.addItem(`p${rng.range(0, 10)}`);
      } else {
        pot.addItem(`f${rng.range(0, 9)}`);
      }
    }
  }

  addEntity(entity) {
    this.entities.add(entity);
    this.entityMap.set(entity.eid, entity);

    this.broadcast(
      new events.Event(
        events.REGION_ENTRANCE,
        `${entity}\n${entity.x} ${entity.y}`,
        entity
      )
    );

    for (let exEnt of this.entities) {
      if (exEnt === entity) {
        continue;
      }

      entity.onEvent(
        new events.Event(
          events.REGION_ENTRANCE,
          `${exEnt}\n${exEnt.x} ${exEnt.y}`
        )
      );
    }

    if (entity instanceof player.Player && this.cleanup) {
      clearTimeout(this.cleanup);
      this.cleanup = null;
    }

  }

  removeEntity(entity) {
    this.entities.delete(entity);
    this.entityMap.delete(entity.eid);
    this.broadcast(
      new events.Event(events.REGION_EXIT, entity.eid, entity)
    );
  }

  getEntity(eid) {
    return this.entityMap.get(eid) || null;
  }

  spawn(entType, x, y) {
    const ent = new entity.VirtualEntity(entType);
    ent.setLocation(this);
    ent.setPosition(x, y);
    this.addEntity(ent);
    return ent.eid;
  }

  getRoot() {
    let parent = this.parentID;
    while (parent !== terrain.WORLD_OVERWORLD && parent !== terrain.WORLD_ETHER) {
      parent = exports.getRegionData(parent)[0];
    }
    return parent;
  }

  toString() {
    return JSON.stringify({
      level: this.terrain.renderDownTilemap(),
      hitmap: this.terrain.hitmap.toArray(),
      tileset: terrain.getTileset(this.getRoot(), this.type),
      can_slide: true,
      h: this.terrain.height,
      w: this.terrain.width,
      x: this.terrain.x,
      y: this.terrain.y,
    });
  }
}


function getRegionID(parent, type, x, y) {
  return `${parent},${type}:${x}:${y}`;
}


exports.getRegion = function(parent, type, x, y) {
  const regionID = getRegionID(parent, type, x, y);

  if (!isValidRegionID(regionID)) {
    console.error(`Invalid region ID requested: ${regionID}`);
    return null;
  }

  if (regionID in regionCache) {
    if (regionCache[regionID].cleanup) {
      clearTimeout(regionCache[regionID].cleanup);
    }
    return regionCache[regionID];
  }

  return new Region(parent, type, x, y);
};

const getRegionData = exports.getRegionData = function(id) {
  const split = id.split(',');
  if (split.length < 2) {
    return DEFAULT_REGION_DATA;
  }

  const regSplit = split[split.length - 1].split(':');
  if (regSplit.length !== 3) {
    return DEFAULT_REGION_DATA;
  }

  const parent = split.slice(0, -1).join(',');
  const x = parseFloat(regSplit[1]);
  const y = parseFloat(regSplit[2]);
  return [parent, regSplit[0], x, y];
};


function isDungeonPos(x, y) {
  return x === 1 && y === 0 || terrain.getCoordOption(x, y, ODDS_DUNGEON);
}

function isTownPos(x, y) {
  return x === 0 && y === 0 || terrain.getCoordOption(x, y, ODDS_TOWN);
}


function isValidRegionID(id) {
  const [parent, type] = getRegionData(id);
  if (parent === terrain.WORLD_OVERWORLD || parent === terrain.WORLD_ETHER) {
    return type === terrain.REGIONTYPE_FIELD;
  }

  if (!isValidRegionID(parent)) {
    return false;
  }

  const [, parentType, parentX, parentY] = getRegionData(parent);
  if (type === terrain.REGIONTYPE_DUNGEON) {
    return parentType === terrain.REGIONTYPE_DUNGEON || parentType === terrain.REGIONTYPE_FIELD && isDungeonPos(parentX, parentY);
  }
  if (type === terrain.REGIONTYPE_FIELD) {
    return parentType === terrain.WORLD_OVERWORLD || parentType === terrain.WORLD_ETHER;
  }

  return true;
}
